lightning/src/routing/router.rs

```diff
@@ -200,100 +378,260 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, targ
+				// if the amount being transferred over this path is lower.
+				// We do this for now, but this is a subject for removal.
+				if let Some(available_value_contribution_msat) = available_liquidity_msat.checked_sub($next_hops_fee_msat) {
+
+					// Routing Fragmentation Mitigation heuristic:
+					//
+					// Routing fragmentation across many payment paths increases the overall routing
+					// fees as you have irreducible routing fees per-link used (`fee_base_msat`).
+					// Taking too many smaller paths also increases the chance of payment failure.
+					// Thus to avoid this effect, we require from our collected links to provide
+					// at least a minimal contribution to the recommended value yet-to-be-fulfilled.
+					//
+					// This requirement is currently 5% of the remaining-to-be-collected value.
+					// This means as we successfully advance in our collection,
+					// the absolute liquidity contribution is lowered,
+					// thus increasing the number of potential channels to be selected.
+
+					// Derive the minimal liquidity contribution with a ratio of 20 (5%, rounded up).
+					let minimal_value_contribution_msat: u64 = (recommended_value_msat - already_collected_value_msat + 19) / 20;
+					// Verify the liquidity offered by this channel complies to the minimal contribution.
+					let contributes_sufficient_value = available_value_contribution_msat >= minimal_value_contribution_msat;
+
+					let value_contribution_msat = cmp::min(available_value_contribution_msat, $next_hops_value_contribution);
+					// Includes paying fees for the use of the following channels.
+					let amount_to_transfer_over_msat: u64 = match value_contribution_msat.checked_add($next_hops_fee_msat) {
+						Some(result) => result,
+						// Can't overflow due to how the values were computed right above.
+						None => unreachable!(),
 					};
-					if old_entry.0 > total_fee {
-						targets.push(new_graph_node);
-						old_entry.0 = total_fee;
-						old_entry.3 = RouteHop {
-							pubkey: $dest_node_id.clone(),
-							node_features: NodeFeatures::empty(),
-							short_channel_id: $chan_id.clone(),
-							channel_features: $chan_features.clone(),
-							fee_msat: new_fee, // This field is ignored on the last-hop anyway
-							cltv_expiry_delta: $directional_info.cltv_expiry_delta as u32,
+
+					// If HTLC minimum is larger than the amount we're going to transfer, we shouldn't
+					// bother considering this channel.
+					// Since we're choosing amount_to_transfer_over_msat as maximum possible, it can
+					// be only reduced later (not increased), so this channel should just be skipped
+					// as not sufficient.
+					// TODO: Explore simply adding fee to hit htlc_minimum_msat
+					if contributes_sufficient_value && amount_to_transfer_over_msat >= $directional_info.htlc_minimum_msat {
+						// Note that low contribution here (limited by available_liquidity_msat)
+						// might violate htlc_minimum_msat on the following hops. To avoid that,
```

Comment:
With this new comment, I think it's easier if you recall the traversal direction here, "following" is a bit ambiguous ("we go payee->our node here").

Given upsetting for htlc_mininum_msat is a side-effect of fee recomputation it should be mentioned otherwise reviewer might be confused. "Recomputing path fees has the side-effect to increase them in case of amount transferred across a hop not matching htlc_minimum_msat"

Reply:
>Given upsetting for htlc_mininum_msat is a side-effect of fee recomputation it should be mentioned otherwise reviewer might be confused.

This is actually *not* a side effect of fee recomputation. It's a side-effect of reducing the value because of low capacity (or htlc_max)

Reply:
Good to me with the new document around `update_value_and_recompute_fees`